### PR TITLE
Add `attributes` property to Device

### LINKED ISSUE #1867: easier way to query device attributes?
I was curious if there is interest in adding a more user-friendly way to retrieve device attributes such as MAX_THREADS_PER_BLOCK, MAX_GRID_DIM_X, etc. In CUDA 9.2/CUDA 10 there are 101 such attributes defined in `cuda.h`. 

It seems that these can already be retrieved via `cupy.cuda.runtime.deviceGetAttribute`, but one needs to know the appropriate enum value for each attribute of interest.

Is there any interest in adding a function that would return the values of all available attributes as a dict or other similar structure that is more user-friendly?

A basic solution I came up with was to copy enum values from `cuda.h` into one of the cython files like `runtime.pxd`.  If there is interest I can make a PR.

e.g.
```cython
cpdef enum:
    CU_DEVICE_ATTRIBUTE_MAX_THREADS_PER_BLOCK = 1              # Maximum number of threads per block
    CU_DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_X = 2                    # Maximum block dimension X
    CU_DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_Y = 3                    # Maximum block dimension Y
    CU_DEVICE_ATTRIBUTE_MAX_BLOCK_DIM_Z = 4                    # Maximum block dimension Z
    CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_X = 5                     # Maximum grid dimension X
    CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_Y = 6                     # Maximum grid dimension Y
    CU_DEVICE_ATTRIBUTE_MAX_GRID_DIM_Z = 7                     # Maximum grid dimension Z
    CU_DEVICE_ATTRIBUTE_MAX_SHARED_MEMORY_PER_BLOCK = 8        # Maximum shared memory available per block in bytes
```
etc.

Then a convenience function similar to the following would create a `dict` with attribute names as the keys.
```python

def get_attributes_dict(device_id):
    """Return a dict containing all device attributes."""
    d = {}
    for k, v in runtime.__dict__.items():
        if 'CU_DEVICE_ATTRIBUTE' in k:
            try:
                name = k.replace('CU_DEVICE_ATTRIBUTE_', '')
                d[name] = runtime.deviceGetAttribute(v, device_id)
            except runtime.CUDARuntimeError:
                pass
    return d
```
(the except case is necessary to skip any attributes that were undefined in older CUDA versions)

The output of the function is a dict such as:
```python
{ 'MAX_THREADS_PER_BLOCK': 1024,
  'MAX_BLOCK_DIM_X': 1024,
  'MAX_BLOCK_DIM_Y': 1024,
  'MAX_BLOCK_DIM_Z': 64,
  'MAX_GRID_DIM_X': 2147483647,
  'MAX_GRID_DIM_Y': 65535,
  'MAX_GRID_DIM_Z': 65535,
  'MAX_SHARED_MEMORY_PER_BLOCK': 49152,
...

### PR DESCRIPTION
**Pull request by grlee77**
This PR adds a memoized `attributes` property to the `Device` class for easy querying of device attributes such as the maximum block size, etc.

The approach follows the suggestions made by @kmaehashi in #1867, although I am not certain if all of the changes are located in the expected files.

closes #1867

### CODE REVIEW

**Comment by hvy** [comment]
@kmaehashi Could you take a look at the changes?

File: cupy/cuda/__init__.py [hunk 1]
```diff
@@ -69,6 +70,20 @@ def get_cuda_path():
     return _cuda_path
 
 
+@util.memoize()
+def get_attributes_dict(device_id):
+    """Return a dict containing all device attributes."""
+    d = {}
+    for k, v in runtime.__dict__.items():
+        if 'cudaDevAttr' in k:
```
Comment (kmaehashi) [inline cupy/cuda/__init__.py]: Use `k.startswith('cudaDevAttr')`.

File: cupy/cuda/__init__.py [hunk 2]
```diff
@@ -69,6 +70,20 @@ def get_cuda_path():
     return _cuda_path
 
 
+@util.memoize()
+def get_attributes_dict(device_id):
```
Comment (kmaehashi) [inline cupy/cuda/__init__.py]: Please move this function to `device.pyx` and name it `_get_attributes` to make it explicitly private.

File: cupy/cuda/runtime.pxd [hunk 3]
```diff
@@ -67,6 +67,106 @@ cpdef enum:
 
     errorMemoryAllocation = 2
 
+    cudaDevAttrMaxThreadsPerBlock             = 1  # Maximum number of threads per block
```
Comment (kmaehashi) [inline cupy/cuda/runtime.pxd]: Could you remove the comment part? (our coding rule (PEP8) requires 80 chars for each line)

File: cupy/cuda/__init__.py [hunk 4]
```diff
@@ -69,6 +70,20 @@ def get_cuda_path():
...
+    """Return a dict containing all device attributes."""
+    d = {}
+    for k, v in runtime.__dict__.items():
+        if 'cudaDevAttr' in k:
+            try:
+                name = k.replace('cudaDevAttr', '')
```
Comment (kmaehashi) [inline cupy/cuda/__init__.py]: Use `k.replace('cudaDevAttr', '', 1)`.

**Comment by kmaehashi** [comment]
Could you try adding documentation for `attributes` property (like https://github.com/cupy/cupy/pull/1869/files#diff-161ddf2c99cd9824466401215e46e033R217), and also adding test code?

**Comment by grlee77** [comment]
Hi @kmaehashi . Thanks for the quick and helpful review!

I have implemented your suggestions and added some tests.

**Comment by kmaehashi** [comment]
LGTM!

Jenkins, test this please.

**Comment by kmaehashi** [comment]
@grlee77 Thanks!

**Pull request merged**